runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -137,31 +131,43 @@ protected OutputT computeNext() {
       // collection (and iterator) is reset between each call to processElement, so the
       // collection only holds the output values for each call to processElement, rather
       // than for the whole partition (which would use too much memory).
-      while (true) {
-        if (outputIterator.hasNext()) {
-          return outputIterator.next();
-        } else if (inputIterator.hasNext()) {
-          clearOutput();
-          // grab the next element and process it.
-          doFnRunner.processElement(inputIterator.next());
-          outputIterator = getOutputIterator();
-        } else if (timerDataIterator.hasNext()) {
+      try {
+        boolean isBundleStarted = false;
+        boolean isBundleFinished = false;
+        while (true) {
+          if (outputIterator.hasNext()) {
+            return outputIterator.next();
+          }
+
+          if (!isBundleStarted) {
+            isBundleStarted = true;
+            // call startBundle() before beginning to process the partition.
+            doFnRunner.startBundle();
+          }
+
           clearOutput();
-          fireTimer(timerDataIterator.next());
-          outputIterator = getOutputIterator();
-        } else {
-          // no more input to consume, but finishBundle can produce more output
-          if (!calledFinish) {
-            clearOutput();
-            calledFinish = true;
-            doFnRunner.finishBundle();
-            // teardown DoFn.
-            DoFnInvokers.invokerFor(doFn).invokeTeardown();
+          if (inputIterator.hasNext()) {
+            // grab the next element and process it.
+            doFnRunner.processElement(inputIterator.next());
+            outputIterator = getOutputIterator();
+          } else if (timerDataIterator.hasNext()) {
+            fireTimer(timerDataIterator.next());
             outputIterator = getOutputIterator();
-            continue; // try to consume outputIterator from start of loop
+          } else {
+            // no more input to consume, but finishBundle can produce more output
+            if (!isBundleFinished) {
+              isBundleFinished = true;
```

Comment:
I think you're missing a call to `clearOutput()`, no?

Reply:
No, notice that the first if guarantees that while we have data it is outputted by the `outputIterater` and then we can confidently clear it because we are constructing new output that will be recaptured (and outputted) if needed by this first if.

Reply:
ah yes, because its called now just once before the `if/elseif/else` block 👍